pkg/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,198 @@
+	return err
+}
+
+func insertMetadataIntoPath(path string) string {
+	pathSplit := strings.Split(path, "/")
+	return strings.Join(append([]string{pathSplit[0], "metadata"}, pathSplit[1:]...), "/")
+}
+
+func insertDataIntoPath(path string) string {
+	pathSplit := strings.Split(path, "/")
+	return strings.Join(append([]string{pathSplit[0], "data"}, pathSplit[1:]...), "/")
+}
+
+func (v *VaultClient) Put(path string, body []byte) error {
+	r := v.Client.NewRequest("PUT", "/v1/"+path)
+	r.BodyBytes = body
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := v.RawRequestWithContext(ctx, r)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+
+	return err
+}
+
+func (v *VaultClient) GetUserByID(id string) (*Entity, error) {
+	var entity Entity
+	return &entity, v.readInto(fmt.Sprintf("identity/entity/id/%s", id), &entity)
+}
+
+func (v *VaultClient) GetGroupNames() ([]string, error) {
+	var result keyResponse
+	if err := v.listInto("identity/group/name", &result); err != nil {
+		return nil, err
+	}
+	return result.Keys, nil
+}
+
+func (v *VaultClient) GetGroupByName(groupName string) (*Group, error) {
+	var group Group
+	return &group, v.readInto(fmt.Sprintf("identity/group/name/%s", groupName), &group)
+}
+
+func (v *VaultClient) GetAllGroups() ([]Group, error) {
+	// The list endpoints return only the id/name
+	names, err := v.GetGroupNames()
+	if err != nil {
+		return nil, err
+	}
+
+	var result []Group
+	for _, name := range names {
+		group, err := v.GetGroupByName(name)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, *group)
+	}
+
+	return result, nil
+}
+
+func (v *VaultClient) GetGroupByID(groupID string) (*Group, error) {
+	var group Group
+	return &group, v.readInto(fmt.Sprintf("identity/group/id/%s", groupID), &group)
+}
+
+func (v *VaultClient) UpdateGroupMembers(groupName string, newMemberIDs []string) error {
+	data := map[string]interface{}{"member_entity_ids": newMemberIDs}
+	_, err := v.Logical().Write(fmt.Sprintf("identity/group/name/%s", groupName), data)
+	return err
+}
+
+func (v *VaultClient) ListAuthMounts() (MountListResponse, error) {
+	var response MountListResponse
+	return response, v.readInto("sys/auth", &response)
+}
+
+func (v *VaultClient) CreateIdentity(name string, policies []string) (*Entity, error) {
+	var entity *Entity
+	return entity, v.writeInto("identity/entity", map[string]interface{}{"name": name, "policies": policies}, &entity)
+}
+
+func (v *VaultClient) listInto(path string, target interface{}) error {
+	raw, err := v.Logical().List(path)
+	if err != nil {
+		return err
+	}
+	// 404 for list means no results: https://github.com/hashicorp/vault/issues/5861
+	if raw == nil || raw.Data == nil {
+		return nil
+	}
+	return v.dataInto(raw.Data, target)
+}
+
+func (v *VaultClient) readInto(path string, target interface{}) error {
+	raw, err := v.Logical().Read(path)
```

Comment:
How are retries, etc handled? Inside of their client?

Reply:
Yeah, they use go-retriable underneath with a couple of custom settings, I thought it was best to just leave the defaults there